Codechange: Use iterators to initialise and allocate rail/road types.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Index-based for-loops to initialise and allocate rail types and road types.

* In initialisation (actually finalisation) two loops occur, one to resolve sprites and another to build a list for sorting.
* In allocation, a `for`-loop is used to find a free slot, and then use it.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

* Perform railtype/roadtype init in a single loop, using iterators instead of indexing.
* For allocation, use std algorithm `std::ranges::find` instead of a `for`-loop.

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

PeterN committed:
Codechange: Make GetRail/RoadTypeInfoIndex() a member method.

This follows a similar pattern used for cargoes, houses and objects.

Also fixes incorrect documentation of GetRoadTypeInfoIndex().

LordAro commented on src/rail_cmd.cpp:
```diff
@@ -129,15 +129,14 @@ static bool CompareRailTypes(const RailType &first, const RailType &second)
-	for (RailType rt = RAILTYPE_BEGIN; rt != RAILTYPE_END; rt++) {
-		if (_railtypes[rt].label == 0) continue;
+	for (auto it = std::begin(_railtypes); it != std::end(_railtypes); ++it) {
+		RailType rt = static_cast<RailType>(it - std::begin(_railtypes));
```
I'm not convinced this is an improvement

PeterN replied:
Fair enough. A different approach...

PeterN committed:
Codechange: Use std algorithm when allocating rail/road type.

Use `std::ranges::find()` instead of a `for`-loop to allocate rail and road types.

PeterN committed:
Codechange: Initialise rail/road types with range-for.

Replace double loop with single loop that performs both parts together.

LordAro commented on src/rail.cpp:
```diff
@@ -19,6 +19,18 @@
+{
+	extern RailTypeInfo _railtypes[RAILTYPE_END];
+	size_t index = this - _railtypes;
+	assert(index < RAILTYPE_END && this == _railtypes + index);
```
I know it was there already, but the second part of this assert seems... unnecessary

Also for the other function
